### ISSUE TITLE
Paradoxical self-reference in endpoints for path constructors

### ISSUE THREAD

**Issue by dwarn** [agda/agda#6336]
Agda 2.6.2.2 accepts the following higher inductive type
```agda
{-# OPTIONS --cubical --safe #-}
open import Agda.Builtin.Cubical.Path
open import Agda.Builtin.Bool

data A (f : Set → Bool) : Set where
  g : Bool → A f
  p : g (f (A f)) ≡ g true
```
but it doesn't actually make sense, since `A f` refers to itself in its own path constructor `p`. This particular example makes Agda *inconsistent* if we assume LEM: we can prove that `A f` is contractible iff `f (A f) = false`, which is impossible if we define `f X := if isContr X then true else false`.

To fix this, Agda should perform some check that the endpoints of path constructors look reasonable (don't refer to the type being constructed in this uncontrollable way).

**Comment by andreasabel** [comment]
@dwarn David, thanks for the report!  
Would be nice to see the exploit spelled out!  (You currently describe it in prose.)
Should we worry about large parameters (like `f`) to HITs in general?

**Comment by nad** [comment]
With `-vtc.pos:50` we get the following output for the code above (if it is saved in `Bug.agda`):
```
computeOccurrences A ConcreteDef ConcreteMode Bug
computing occurrences in  (_ : Bool)
computing occurrences in  []
computing occurrences in  (i : Agda.Primitive.Cubical.I)
computing occurrences in  []
Occurrences in A:
  ADef Agda.Builtin.Bool.Bool: 1 occurrences
  ADef Agda.Primitive.Cubical.I: 1 occurrences
computeOccurrences g ConcreteDef ConcreteMode Bug
Occurrences in g:
computeOccurrences p ConcreteDef ConcreteMode Bug
Occurrences in p:
Positivity graph: N=0 E=0
positivity graph for A g p
Positivity graph (completed): E=0
transitive closure of positivity graph for [A, g, p]
checking args of A
args of A = [-[_]->]
0 trivial sccs 0 non-trivial sccs with lengths []
  sccs = []
Checking positivity of A
```
Perhaps it suffices to revise the following code in some way:

https://github.com/agda/agda/blob/7de89ad28483b0b9df09b1651700c1472c653726/src/full/Agda/TypeChecking/Positivity.hs#L520-L573

If this is not too hard, then I suggest that we include a fix in Agda 2.6.3.

**Comment by nad** [comment]
> ```
> computing occurrences in  (i : Agda.Primitive.Cubical.I)
> ```

Is the type of the constructor expanded to `I → A f` before positivity is checked? I guess that this is due to the use of `telViewPath`:

https://github.com/agda/agda/blob/7de89ad28483b0b9df09b1651700c1472c653726/src/full/Agda/TypeChecking/Positivity.hs#L535-L539

If `telViewPath` is replaced by `telView`, then the code in the OP triggers an internal error:

https://github.com/agda/agda/blob/7de89ad28483b0b9df09b1651700c1472c653726/src/full/Agda/TypeChecking/Positivity.hs#L557-L562

Does it suffice to add a case for paths here? If the check `q == q'` is removed, then we get false positives in `test/Succeed/HITs.agda`. One example:

https://github.com/agda/agda/blob/7de89ad28483b0b9df09b1651700c1472c653726/test/Succeed/HITs.agda#L19-L21

```
∥_∥ is not strictly positive, because it occurs
in an index of the target type of the constructor squash
in the definition of ∥_∥.
```

If the target of a constructor is an iterated path type, which occurrences should be included? If the target is something like `Path (Path (D ps is) e₁ e₂) e₃ e₄`, then I guess one should include occurrences in the indices `is` as well as in the expressions `e₁`, `e₂`, `e₃` and `e₄`, but perhaps not occurrences in `ps`.

**Comment by andreasabel** [comment]
Indeed, by using `telViewPath`, we are simply dropping the boundaries, so `g (f (A f)) ≡ g true` becomes just `I -> A f` in the eyes of the positivity checker.
We might want to also retrieve the boundaries, e.g. by using `telViewUpToPathBoundary' (-1)` instead, and subject them to checks.  A simple check would assert mixed-variance to any occurrence of the mutually defined data types here, failing the positivity check.  However, this might be ruling out benign cases such as:
```agda
data D : Set where
  point : D
  loop : point == id D point
```
(Example by @dwarn.)

**Comment by andreasabel** [comment]
I am a bit pessimistic that this can go into 2.6.3.  I'd rather release the status quo as known bug rather than adding extra checks that might be too strict just before the release. Also it is not a recent regression, the OP already passes with 2.6.0.

**Comment by nad** [comment]
> We might want to also retrieve the boundaries, e.g. by using `telViewUpToPathBoundary' (-1)` instead, and subject them to checks. A simple check would assert mixed-variance to any occurrence of the mutually defined data types here, failing the positivity check.

I implemented this. Then we get something like the following error message for the code in the OP:
```
A is not strictly positive, because it occurs
in an argument of a bound variable
in an endpoint of the target of the higher constructor p
in the definition of A.
```

> However, this might be ruling out benign cases such as:
> 
> ```agda
> data D : Set where
>   point : D
>   loop : point == id D point
> ```

This code is still accepted. However, my rendering of the definition of K(G, 1) from "Eilenberg-MacLane Spaces in Homotopy Type Theory" by Licata and Finster is rejected because it uses transitivity in one endpoint, and the transitivity proof takes the type that is defined as one argument. See `test/Fail/Issue6017.agda` for a similar example. We get this problem also if `mixed` is replaced by `keepGoing` for path endpoints in `computeEdges`.

Can anyone suggest an approach that would work?

> Should we worry about large parameters (like `f`) to HITs in general?

Including `A : Set`?

**Comment by andreasabel** [comment]
@nad Maybe shipping your solution could be fine since it can be worked around by switching off the positivity checker.

**Comment by jespercockx** [comment]
@nad I do not understand why this is on 2.6.3.

**Comment by nad** [comment]
Because it is a known inconsistency.

**Comment by nad** [comment]
> @nad Maybe shipping your solution could be fine since it can be worked around by switching off the positivity checker.

I could make all HITs in the Agda code that I maintain work in the following way:
* I replaced `mixed` with `keepGoing`. This means that an occurrence in a "strict" position in an argument is not automatically treated as "bad".
* I made some arguments of `primHComp` "strictly positive", the type argument (`A`) as well as the last two arguments:
   https://github.com/agda/agda/blob/48fcdd5c5fbbc4c25ff9d066445fcb9c419c51e6/src/data/lib/prim/Agda/Primitive/Cubical.agda#L72
  Note that this can make Agda accept some code that was previously rejected. I don't know if this change is OK. If it does make sense, then one could consider declaring more argument occurrences for `primHComp`, and argument occurrences for other cubical primitives.
* I inlined some definitions in my Agda code.

Do these changes make sense? (See pull request #6344 for details, along with a comment related to the cubical library.)

**Comment by jespercockx** [comment]
> Because it is a known inconsistency.

We usually don't require all inconsistencies to be fixed before a release, unless they are regressions (see e.g. sized types)

**Comment by nad** [comment]
> We usually don't require all inconsistencies to be fixed before a release, unless they are regressions (see e.g. sized types)

We fixed the sized types problem by disallowing sized types when `--safe` is used. We could fix this problem by disallowing HITs when `--safe` is used. However, I would prefer something less disruptive.

**Comment by MatthiasHu** [comment]
@dwarn and I just worked out the inconsistency with propositional resizing using the cubical library.

```agda
{-# OPTIONS --cubical --safe #-}

open import Cubical.Foundations.Prelude
open import Cubical.Foundations.HLevels
open import Cubical.Foundations.Equiv
open import Cubical.Foundations.Isomorphism
open import Cubical.Foundations.Structure

open import Cubical.Data.Unit
import Cubical.Data.Empty as ⊥

open import Cubical.Functions.Logic

open import Cubical.HITs.PropositionalTruncation as PT

-- Let's assume propositional resizing, in the form that the type of small propositions is equivalent to a small type.
module _
  (Ω : Type ℓ-zero)
  (Ω≃hProp : Ω ≃ hProp ℓ-zero)
  where

  ⊤Ω : Ω
  ⊤Ω = invEq Ω≃hProp ⊤

  ⊥Ω : Ω
  ⊥Ω = invEq Ω≃hProp ⊥

  data A : Set where
    f : Ω → A
    p : f ⊤Ω ≡ f (invEq Ω≃hProp (∥ f ⊤Ω ≡ f ⊥Ω ∥₁ , isPropPropTrunc))

  module _
    (f⊤≡f⊥ : f ⊤Ω ≡ f ⊥Ω)
    where

    lemma : ⊤Ω ≡ invEq Ω≃hProp (∥ f ⊤Ω ≡ f ⊥Ω ∥₁ , isPropPropTrunc)
    lemma = cong (invEq Ω≃hProp) (⇔toPath (λ _ → ∣ f⊤≡f⊥ ∣₁) (λ _ → tt*))

    f' : A → Ω
    f' (f x) = x
    f' (p i) = lemma i

    ⊤≡⊥ : ⊤ ≡ ⊥
    ⊤≡⊥ = isoInvInjective (equivToIso Ω≃hProp) _ _ (cong f' f⊤≡f⊥)

    f⊤≢f⊥ : ⟨ ⊥ ⟩
    f⊤≢f⊥ = subst ⟨_⟩ ⊤≡⊥ tt*

  ¬∥f⊤≡f⊥∥ : ∥ f ⊤Ω ≡ f ⊥Ω ∥₁ → ⟨ ⊥ ⟩
  ¬∥f⊤≡f⊥∥ = PT.rec (snd ⊥) f⊤≢f⊥

  bad : ⟨ ⊥ ⟩
  bad = f⊤≢f⊥ (p ∙ (cong f (cong (invEq Ω≃hProp) (⇔toPath ¬∥f⊤≡f⊥∥ ⊥.rec))))
```

**Comment by nad** [comment]
> We usually don't require all inconsistencies to be fixed before a release, unless they are regressions (see e.g. sized types)

I noticed that there is another inconsistency related to Cubical Agda which has not been fixed (#5910), and that one does not even require any assumptions. There is also a third one which makes use of a classical axiom (#4966).

It is not quite clear how to fix the current issue: as I noted above I don't know if my change is OK, and no one has commented on this. Thus I do not think that we should fix this before the release.

**Comment by plt-amy** [comment]
@nad I think your fix in #6344 is alright, but as mentioned there I'm concerned about the issue with variables you observed; I'll try to investigate later. However, even if we merge that, I'd like for this issue to stay open.

Cubical Agda is, IMO, too lax with what is allowed in boundary terms, contrary to the schema for HITs from @ecavallo's work (the only one I'm aware of! See e.g. [here](https://staff.math.su.se/evan.cavallo/works/popl19.pdf)​), which takes the opposite approach and is fairly _restrictive_. We need to have a _proper_ think about what should, and should not, be allowed in the boundaries of constructors: according to that paper, for example, the type being defined should _not_.

That would break some code, like Andreas' benign example, but also "serious" code which defines HITs with path composition in their boundaries, but I'm actually half-convinced that'd be a good thing to break... Can anyone in this thread figure out the solution to this interaction point, without loading up Agda?

```agda
data X : Set where
  b : X
  l : b ≡ b
  p : l ∙ l ≡ l

recX : {B : Set} (mb : B) (ml : mb ≡ mb) → (mp : {!   !}) → X → B
recX mb ml mp b = mb
recX mb ml mp (l i) = ml i
recX mb ml mp (p i j) = mp i j
```

The solution is `(λ i → comp (λ i → B) (λ { j (i = i0) → mb ; j (i = i1) → ml j }) (ml i)) ≡ ml`, which I'm sure everyone was able to guess. My point is that having `hcomp` in boundaries has _very_ unexpected results, and **I'm not sure** whether we should let people _accidentally_ generate complexity like this. Evan's schema would disallow the type above, since it requires `X` to be in scope for the application of `_∙_`. It would _not_ disallow this:

```agda
data X : Set where
  b : X
  l : b ≡ b
  p : (λ i → hcomp (λ { j (i = i0) → b ; j (i = i1) → l j }) (l i)) ≡ l
```

which is much clearer about the fact that it'll ruin your use-sites. Ah, I should add that a naïve implementation of this in Agda would forbid the code above _as well_, so we have to be careful to make sure that the type being defined can appear only in `hcomp` (and `transp`).
